infra/utils.sh

```diff
@@ -0,0 +1,173 @@
+    if ! bq --project "${project}" ls "${dataset_name}" >/dev/null 2>&1; then
+        bq mk --project_id="${project}" --location="${region}" "${dataset_name}"
+    else
+        printf "Found existing %s dataset.\n" "${dataset_name}"
+    fi
+}
+
+function create-cluster() {
+    gcloud beta container clusters create "${NAME}" \
+        --project "${PROJECT}" \
+        --region "${REGION}" \
+        --machine-type "${MACHINE_TYPE}" \
+        --disk-type "${DISK_TYPE}" \
+        --cluster-version "${CLUSTER_VERSION}" \
+        --num-nodes "${NUM_NODES}" \
+        --enable-network-policy \
+        --scopes "${SCOPES}" \
+        --network "${NAME}" \
+        --subnetwork "${NAME}-custom-subnet" \
+        --enable-ip-alias \
+        --enable-private-nodes \
+        --enable-master-authorized-networks \
+        --master-ipv4-cidr "${MASTER_IPV4_CIDR}" \
+        --master-authorized-networks "${MASTER_IP_RANGE}" \
+        --max-pods-per-node "${MAX_PODS_PER_NODE}" \
+        --no-issue-client-certificate \
+        --metadata disable-legacy-endpoints=true \
+        --service-account "$(svc_acct_for "${PROJECT}" "${SVC_ACCOUNT}")" \
+        --security-group "gke-security-groups@kubernetes.io" \
+        --enable-stackdriver-kubernetes \
+        --resource-usage-bigquery-dataset "${DATASET_NAME}" \
+        --quiet
+
+    if ! gcloud beta container clusters describe test-cluster --format="value(resourceUsageExportConfig)" >/dev/null 2>&1; then
+        gcloud beta container clusters update "${NAME}" --region "${REGION}" --resource-usage-bigquery-dataset "${DATASET_NAME}"
+    fi
+}
+
+# Sets up Cloud NAT for the network.
+# Assumed vars:
+#   NETWORK_PROJECT
+#   REGION
+#   NETWORK
+function create-cloud-nat-router() {
+    if [ $# -lt 1 ] && [ $# -gt 2 ] && [ -z "$1" ]; then
+        printf "create-cloud-nat-router(project, network, region) requires 2 or 3 arguments" >&2
+        return 1
+    fi
+    local -r project="${1}"
+    local -r network="${2}"
+    local router_name="$network-nat-router"
+    local nat_name="$network-nat-config"
+    local region=${region:=us-central1}
+
+    if ! gcloud --project "${project}" compute routers describe "${router_name}" &>/dev/null; then
+        gcloud --project "${project}" compute routers create "${router_name}" \
+            --region "${region}" \
+            --network "${network}"
+    else
+        printf "Found existing CloudNAT router : %s.\n" "${router_name}"
+    fi
+
+    if ! gcloud --project "${project}" compute routers nats describe "${nat_name}" --router "${router_name}" &>/dev/null; then
+        gcloud --project "${project}" compute routers nats create "${nat_name}" \
+            --router-region "${region}" \
+            --router "${router_name}" \
+            --nat-all-subnet-ip-ranges \
+            --auto-allocate-nat-external-ips
+    else
+        printf "Found existing NAT route : %s.\n" "${nat_name}"
+    fi
+}
+
+function create-network() {
```

Comment:
Why do we need custom networks, and a cloud nat? Is there a need for these to be private IP clusters?

Reply:
A custom network allow me set a subnet with a specific network range. We could have a subnet per cluster with all subnets associated to the same network. 
With a private cluster, the nodes don't have outbound Internet access because they do not have external IP addresses. The Cloud NAT will used to provide outbound Internet access, in case we need to pull public docker images.

Reply:
Right, but why is it that we need a specific network range? Why do we need private IPs? Public IPs should be fine.

Reply:
There is not a specific need. Only control how the pool of IP address will be used for this specific cluster. 
Having private IPs for the nodes reduce the surface attack and prevent exceeding the quota of Public IPs. The default quota (if I'm correct) of public addresses for this project is in `us-central1` for example : 
```console
gcloud compute regions describe us-central1 --format='value(quotas[2])'
limit=21.0;metric=STATIC_ADDRESSES;usage=0.0
```
@thockin can you confirm ? 
@spiffxp WDYT ?

Reply:
There's no reason NOT to use a private cluster.  It doesn't limit LBs or anything, just changes how we access the master.

Also, this should probably be ensure_network, and auto mode should be fine.  That way we don't have to allocate a subnet manually

Reply:
or we use GKE's --create-subnetwork flag

Reply:
It's extra, non-default complexity. If we don't run it, then we don't need to worry about a cloud NAT or anything like that. I don't believe any of our current clusters are private.

Reply:
Use this flag implies create a subnetwork in the `default` network. If you still choose to create private clusters, we just need to create only one CloudNAT routler for the `default` network.